Add event for adding importer options + mbsubmit plugin

diego-plan9 opened the pull request:
As part of the follow-up on #1689, here is my stab at attempting to add a new event that allows plugin to append options (choices) to the importer prompts.

Currently, it is fired during the main `while` loop at `TerminalImportSession.choose_candidate()`, with `session` and `task` as its parameters. It assumes that the plugins are responsible for checking the conditions for their desired extra choices (ie. if the plugin wants to add an option that is only appended when there are no candidates, it should do the pertinent checks on `task.candidates` inside the function that receives the event).

I have toyed with the alternative of firing it just before the actual calls to `ui.input_options()` inside `commands.choose_candidate()`, and using more fine-grained attributes on the event (singleton, candidates, etc). In the end I discarded that route as, while it might be a bit more friendly for the plugins, I felt it restricted flexibility significantly (unless coming up with a comprehensive list of attributes potential plugins might be interested in).

A sample `mbsubmit` plugin is included as well, which basically serves as an example of the current version of the event (and includes some extra verbose comments and an extra callback that will be removed in the final version). 

As usual, please let me know if this is the right track! I have not started formal testing yet, so please be wary of potential mix ups and so on.

diego-plan9 commented on beets/ui/commands.py:
```diff
@@ -39,6 +42,7 @@
 from beets.util.confit import _package_path
 
 VARIOUS_ARTISTS = u'Various Artists'
+ExtraChoice = namedtuple('ExtraChoice', ['plugin', 'id', 'long', 'callback'])
```
This is basically for readability. The `plugin` field is not currently used - I left it there just in case the actual plugin is needed for something (for example, getting its name for displaying the user which choices are standard, and which choices are appended by that plugin).

sampsyo replied:
Great! I like namedtuples; they really can improve readability.

What do you think about calling this something more generic, like `PromptChoice`, `UIChoice`, or just plain `Choice`? That way, we could (someday) use the same infrastructure for built-in choices as plugins use for "extra" choices. As much as is reasonable, it's nice to use common tooling for both plugin and non-plugin functionality so we can avoid duplication.

diego-plan9 replied:
Absolutely, the same thought did cross my mind while revising the code (in particular, the `if`s that compare the `input_options()` return value with raw characters struck me as a bit out of place), and paving the way for handling all choices in the same way regardless of their origin sounds great.

diego-plan9 commented on beets/ui/commands.py:
```diff
@@ -489,6 +494,13 @@ def choose_candidate(candidates, singleton, rec, cur_artist=None,
+    # TODO: move "abort" choice always to the end?
+    # TODO: add plugin name info on choices ("(mb) Print tracks")?
+    extra_opts = tuple([c.long for c in extra_choices])
+    extra_actions = {c.long.strip(lowercase + whitespace)[0].lower(): c.id
```
This one is a bit ugly, but its purpose is to find out the character based on the "long" description of the choice ("do someThing" -> "t").

sampsyo replied:
Got it; this makes sense. Either way (implicit or explicit) seems fine—if we go implicit, we'll just have to make it clear in the docs that the capitalized letter is how you control your action letter.

diego-plan9 replied:
Great! I was holding off the writing documentation until making sure the changes were on the right track, but will add it soon. On a second look, I will probably go for the explicit route - it should be rather lightweight for the plugin writers, and will probably make things a bit more clear and less prone to surprises.

diego-plan9 commented on beets/ui/commands.py:
```diff
@@ -684,10 +704,18 @@ def choose_match(self, task):
+            # TODO: check for conflicts in choices (duplicated letters, etc)
+            # TODO: provide a way to override summary_judment and/or non-timid
+            # mode? Might be useful for some plugins
+            extra_choices = list(chain(*plugins.send('before_choose_candidate',
```
The list flattening is to handle extra choices received from various plugins.

As hinted on the TODO, it might be useful to do an extra-pass for solving potential problems with the extra choices  (two plugins picking the same `id` for their choices, the same short letter, etc).

sampsyo replied:
Yes, just crashing when there's a letter conflict would probably be the best policy.

I think we can defer other option-related functionality (like `summary_judgment`) to another iteration. It's a great point, but this will be fine for a first version of the hook.

sampsyo replied:
Maybe a helper function like `get_plugin_options` to wrap up this list-flattening and such would aid readability?

diego-plan9 replied:
Fully agree (once again)! It should not be too complex to provide a simple way to try solve letter conflicts (loop on the characters of the long string of the choices until a "free" character is found), but seems like just crashing with an informative description should be enough until more plugins start using the event.

diego-plan9 commented on beets/ui/commands.py:
```diff
@@ -708,6 +736,15 @@ def choose_match(self, task):
+                if post_choice in (importer.action.SKIP, importer.action.ASIS,
+                                   importer.action.TRACKS,
+                                   importer.action.ALBUMS):
+                    return post_choice
```
This is illustrated on `mbsubmit.print_tracks_and_skip()`: if a callback returns a value, the main loop returns that value as well. The idea and main benefit is to save the user an extra prompt - not sure if it introduces more problems/lack of intuitiveness than it solves, as some of the actions (`MANUAL` and `MANUAL_ID`) do not really exit the loop, they are currently not really handled properly.

sampsyo replied:
Sure; this makes sense. And not permitting `MANUAL` and `MANUAL_ID` seems OK, since these are really part of the UI more than they are part of the importer. In fact, an argument could be made for removing them from the `importer.action` enum and making them the sole purview of the `TerminalImportSession` component.

diego-plan9 replied:
Great, I'll just make sure to mention it on the documentation as it might not be evident for the plugin developers.

sampsyo commented on beets/ui/commands.py:
```diff
@@ -489,6 +494,13 @@ def choose_candidate(candidates, singleton, rec, cur_artist=None,
+    # Build helper variables for extra choices.
+    # TODO: move "abort" choice always to the end?
+    # TODO: add plugin name info on choices ("(mb) Print tracks")?
+    extra_opts = tuple([c.long for c in extra_choices])
```
FWIW, `tuple(c.long for c in extra_choices)` without the square brackets will work (using a "generator expression").

diego-plan9 replied:
Thanks for the tip, as usual!

sampsyo commented:
This is looking *awesome*. The changes to the importer workflow are not too invasive, and the functionality will be useful. The edit plugin, for example, should be able to use the same hooks. Woohoo!

diego-plan9 commented:
Thanks, I'm glad the changes are looking good! One of my goals was indeed to try to come up with a minimal set of changes that were able to provide the desired "add choices to the prompt" feature while being as less intrusive as possible.

Out of curiosity, do you recall off the top of your head if any other current plugins would potentially benefit from this functionality? I can investigate further and (if needed and desirable) do a "plugin update" round once this pull request is completed.

diego-plan9 commented:
Added the changes suggested on the line notes:
* `ExtraChoice` is now `PromptChoice`, and has a `short` attribute
* event sending, flattening, and conflict resolution moved to `_get_plugin_options()`
* removed extra-comments and TODOs from the code and plugin and placed them in `plugins.rst`

Still pending on my end: add the hook also for singletons (`choose_item()`), add tests.

diego-plan9 commented on beets/ui/commands.py:
```diff
@@ -801,6 +830,53 @@ def should_resume(self, path):
+                   for letter in duplicates}
+            conflict_msg = '; '.join("'%s': %s" % (k, ', '.join(v))
+                                     for (k, v) in dup.iteritems())
+            raise ValueError("Prompt options have the same short letter\n%s" %
```
Might have gone overboard on this one - the end goal is to come up with an informative message, such as:
```
ValueError: Prompt options have the same short letter
't': TerminalImportSession:"as Tracks", MBSubmitPlugin:"print Tracks"
```

sampsyo replied:
This seems totally reasonable. My only question is whether we really need to have the `plugin` member for `PromptChoice`. It helps a bit with debuggability here, but plugin authors should be able to tell when they are responsible for creating a conflict. And if we want to have a uniform interface with built-in choices, then it will become decreasingly relevant. For example, "TerminalImportSession" is already a little weird in a diagnostic message.

sampsyo replied:
One other change to consider: Instead of throwing an exception here, we could just log an error and remove one of the duplicates. Then we would at least avoid interrupting import sessions if two plugin developers happen to overlap.

diego-plan9 replied:
> One other change to consider: Instead of throwing an exception here, we could just log an error and remove one of the duplicates. Then we would at least avoid interrupting import sessions if two plugin developers happen to overlap.

Sure, I'd be happy to try that approach. Re-assigning the letters of the rest of the duplicates could also be an option, and probably not too costly (loop trough the long description of the choices looking for "free" letters), and I'd be happy to give it a try if it sounds like a good idea.

sampsyo replied:
Yeah, that would definitely work—but for the sake of simplicity, I would lean toward just "giving up" on conflict for the first iteration. If conflicts become a real problem, we can make the policy smarter.

sampsyo commented:
> Out of curiosity, do you recall off the top of your head if any other current plugins would potentially benefit from this functionality? I can investigate further and (if needed and desirable) do a "plugin update" round once this pull request is completed.

Good question. I think most of the use cases that have been discussed are plugins that don't exist yet because they would require some kind of extra prompt. Here are a few issues on the topic of new prompts that could (ahem) prompt new plugins: #1678, #1064, #396, #112.

sampsyo commented on beets/ui/commands.py:
```diff
@@ -471,7 +475,8 @@ def _summary_judment(rec):
-                     cur_album=None, item=None, itemcount=None):
+                     cur_album=None, item=None, itemcount=None,
+                     extra_choices=[]):
     """Given a sorted list of candidates, ask the user for a selection
```
The docstring could use a note on the type and purpose of `extra_choices`.

sampsyo commented on beets/ui/commands.py:
```diff
@@ -708,6 +729,14 @@ def choose_match(self, task):
+                if post_choice in (importer.action.SKIP,
+                                   importer.action.ASIS,
+                                   importer.action.TRACKS,
+                                   importer.action.ALBUMS):
```
Maybe `isinstance(post_choice, importer.action)` would be fine—as long as we kindly ask people not to use `MANUAL` and `MANUAL_ID`?

sampsyo commented on beets/ui/commands.py:
```diff
@@ -801,6 +830,53 @@ def should_resume(self, path):
                            "was interrupted. Resume (Y/n)?"
                            .format(displayable_path(path)))
 
+    def _get_plugin_options(self, task):
```
Perhaps `_get_plugin_choices` would be a more consistent name?

sampsyo replied:
Of course, I now notice that I suggested the name with `choices`. :flushed: Sorry!

diego-plan9 replied:
No worries, I'll update the name!

Also, I kind of defaulted to using "*choice*" for the concept of "*an item the user can select during the prompt*", but it might be a language issue. If you feel that it is more natural to use "*option*" for that concept, please let me know so I can make a pass at the code and documentation for trying to make everything more consistent.

sampsyo replied:
Cool! *Choice* seems like a good name, since it avoids overlap with "configuration option" or "command-line option," as in a flag. I can't think of anything better at the moment.

sampsyo commented on docs/dev/plugins.rst:
```diff
@@ -528,3 +533,63 @@ command and an import stage, but the command needs to print more messages than
+will be passed to the main loop upon the callback has been processed. Note that
+``action.MANUAL`` and ``action.MANUAL_ID`` will have no effect even if
+returned by the callback, due to the current arquitecture of the import
+process.
```
This documentation is perfect! It makes the purpose very clear, and the example is very helpful.

sampsyo commented:
One other design comment crossed my mind: do we really need the `id` field on PromptChoice? If they are uniquely identifiable by the `short` letter, then we can use that as an identifier. And it doesn't seem like the plugin has a need for the `id` either—it knows what was selected via the callback, not the id. Could we get away without it?

Kraymer commented on docs/dev/plugins.rst:
```diff
@@ -528,3 +533,63 @@ command and an import stage, but the command needs to print more messages than
+be performed by returning one of the values from ``importer.action``, which
+will be passed to the main loop upon the callback has been processed. Note that
+``action.MANUAL`` and ``action.MANUAL_ID`` will have no effect even if
+returned by the callback, due to the current arquitecture of the import
```
arquitecture => architecture

diego-plan9 replied:
Thanks for the correction! As with the rest of my documentation commits, I do welcome another pair of eyes reading through the changes and making sure no language issues slip through.

diego-plan9 commented:
> One other design comment crossed my mind: do we really need the id field on PromptChoice? If they are uniquely identifiable by the short letter, then we can use that as an identifier. And it doesn't seem like the plugin has a need for the id either—it knows what was selected via the callback, not the id. Could we get away without it?

Yes, the `id` field was added as a way to mimic the "static" identifiers on `importer.action` (SKIP, ASIS, etc), but we definitely can get away without it - the only requirement is that the comparisons that use `id` currently use something that is guaranteed to be unique, which can indeed simply be the short letters.

Also, related to https://github.com/sampsyo/beets/pull/1758#discussion_r47688551 , `plugin` seems like it could safely be removed as well - you are right in that we can hopefully trust the plugin authors to be able to solve the possible conflicts, making displaying the plugin name in the assertion kind of irrelevant. I'll strip both fields (`id` and `plugin`) on the next iteration!

sampsyo commented:
Awesome! Thanks for working to make the interface simple—this will be great for the long-term evolution of the feature.

diego-plan9 commented:
Pushed a handful of commits that hopefully address the previous issues (simplification of `PromptChoice`, docstring clarifications, function renaming, docs fixes, and singleton handling), along with some unit tests that are also aimed to demonstrate and test basic functionality.

Apart from any other issues or adjustments that might be discovered during the next code review, hopefully the only missing piece of the puzzle to finally complete #1689 would be to work on the `mbsubmit` plugin itself (I expect it to be rather tiny, mainly discussing some things such as when it should be applied and maybe add some configuration options) - perhaps it should be removed from this pull request onto a new one, in order not to pollute this one too much?

diego-plan9 commented on beets/ui/commands.py:
```diff
@@ -801,6 +845,48 @@ def should_resume(self, path):
+                    log.warn(u"Prompt choice '{0}' removed due to conflict "
+                             u"with '{1}' (short letter: '{2}')",
+                             c.long, dup_choices[0].long, c.short)
+                    extra_choices.remove(c)
```
The conflict resolution makes sure that no "core" choices are removed, as they are at the beginning of `all_choices`. No real effort is made on deciding which choice to keep if they are all from plugins - "first come, first served" of sorts.

diego-plan9 commented on test/test_plugins.py:
```diff
@@ -401,6 +402,155 @@ def dummy9(self, **kwargs):
+        self.teardown_plugin_loader()
+        self.teardown_beets()
+
+    def test_plugin_choices_in_ui_input_options_album(self):
```
This one and `test_plugin_choices_in_ui_input_options_singleton` make assertions on the parameters passed to `ui.input_options` as a way to verify that the options are indeed presented to the user - I felt this was cleaner than directly comparing `stdout`.

diego-plan9 commented on test/test_plugins.py:
```diff
@@ -401,6 +402,155 @@ def dummy9(self, **kwargs):
+
+        # DummyPlugin.foo() should be called once
+        with patch.object(DummyPlugin, 'foo', autospec=True) as mock_foo:
+            with helper.control_stdin('\n'.join(['f', 's'])):
```
This is a bit hackish, but `TestTerminalImportSession._add_choice_input` has some checks in place that prevent the use of `self.importer.add_choice('f')` directly.

sampsyo commented:
Awesome!! Thanks for your patience while I was on a brief vacation. :ski:

This is looking perfect. I'll merge it now—let's consider the `mbsubmit` plugin in its current version an "illustrative prototype" rather than a complete plugin. Since it's undocumented, we won't bother any users by adding the initial version to the repo now.

Woohoo! :sparkles:

diego-plan9 commented:
I'm glad the changes have been merged, and, of course, I'm fully aware that these weeks tend to be a bit slow for usually (and hopefully!) good reasons! I'll be working on cleaning up and completing the `mbsubmit` plugin in the hopes of finally closing the initial issue.
